### PR TITLE
[Clang] Allow raw string literals in C as an extension

### LINKED ISSUE #85703: GCC compatibility: Raw strings in C mode
GCC in -xc -std=gnuNN mode accepts C++11 raw strings. https://godbolt.org/z/qoM55qqfz
```c++
// Type your code here, or load an example.
const char * square(void) {
    return R"(a"b"c)";
}
```
Perhaps Clang should too?

(Ideally, I'd like to see that feature present in C3x or something, but that's far off. It would technically break existing source code that #defines R to [something funny](https://godbolt.org/z/EWYa3hvhh).)

### PR DESCRIPTION
**Pull request by Sirraide**
This is a tentative implementation of support for raw string literals in C following the discussion on #85703. 

GCC supports raw string literals in C in `-gnuXY` mode. This pr both enables raw string literals in `-gnuXY` mode in C and adds a `-f[no-]raw-string-literals` flag to override this beheviour. There are a few questions I still have though:

1. GCC does not seem to support raw string literals in C++ before C++11, even if e.g. `-std=gnu++03` is passed. Should we follow this behaviour or should we enable raw string literals in earlier C++ language modes as well if `-gnu++XY` is passed? `-fraw-string-literals` currently makes it possible to enable them in e.g. C++03.
2. `-fno-raw-string-literals` allows users to *disable* raw string literals in `-gnuXY` mode. I thought it might be useful to have this, but do we want it?
3. The implementation of this currently adds a `RawStringLiterals` option to the LangOpts; `-f[no-]raw-string-literals` overrides the default value for it which depends on the language standard. As a consequence, passing e.g. `-std=c++11 -fno-raw-string-literals` will *disable* raw string literals even though we’re in C++11 mode. Do we want to allow this or should we just ignore `-f[no-]raw-string-literals` if we’re in C++11 or later?
4. This probably deserves a note in `LanguageExtensions.rst`, but I’m not exactly sure where.
5. Should we add a flag for this to `__has_feature`/`__has_extension`?

### CODE REVIEW

**Review by AaronBallman — COMMENTED** [review]
Thank you for working on this! Broadly speaking, I think the idea makes a lot of sense.

>  GCC does not seem to support raw string literals in C++ before C++11, even if e.g. -std=gnu++03 is passed. Should we follow this behaviour or should we enable raw string literals in earlier C++ language modes as well if -gnu++XY is passed? -fraw-string-literals currently makes it possible to enable them in e.g. C++03.

I think we should follow that behavior; because `R` can be a valid macro identifier, being conservative is defensible.

>  -fno-raw-string-literals allows users to disable raw string literals in -gnuXY mode. I thought it might be useful to have this, but do we want it?

I think it's reasonable to have it, but I don't think we should allow it for C++11 and later modes unless there's some rationale I'm missing. (I don't think we want to let users disable language features in standards modes where the feature is standardized without some sort of reasonable justification.)

> The implementation of this currently adds a RawStringLiterals option to the LangOpts; -f[no-]raw-string-literals overrides the default value for it which depends on the language standard. As a consequence, passing e.g. -std=c++11 -fno-raw-string-literals will disable raw string literals even though we’re in C++11 mode. Do we want to allow this or should we just ignore -f[no-]raw-string-literals if we’re in C++11 or later?

I think we should either ignore or diagnose it in C++11 or later.

>  This probably deserves a note in LanguageExtensions.rst, but I’m not exactly sure where.

It definitely should be noted in there; I would probably recommend https://clang.llvm.org/docs/LanguageExtensions.html#c-11-raw-string-literals for the C++ side of things and then something similar for C around where we document those.

> Should we add a flag for this to __has_feature/__has_extension?

Yes, but it's a fun question as to which one. We currently use `__has_feature` for it in C++:
```
FEATURE(cxx_raw_string_literals, LangOpts.CPlusPlus11)
```
and it seems like it would make sense to continue to do so for C++. But this isn't a language feature of C, so `__has_extension` makes sense there. But that's confusing because then we've got both, so I'm not entirely certain that's the right approach. Perhaps using `__has_feature` for both C and C++ makes the most sense?

File: clang/include/clang/Basic/LangStandard.h
Comment (Sirraide) [reply clang/include/clang/Basic/LangStandard.h:136]: Update: This seems to be intended, so I suppose we should only support them in `-gnu99` mode and later.

Comment (AaronBallman) [reply clang/include/clang/Basic/LangStandard.h:136]: okay, so my guess was correct. It's still odd though because that same logic applies to C99 (it's "old" and the behavior can still change). But yeah, we should follow the same behavior as GCC in this case (less confusion for users).

**Comment by Sirraide** [comment]
> I don't think we should allow it for C++11 and later modes

To clarify, should we allow enabling them in e.g. `c++03` mode if `-fraw-string-literals` is passed? I don’t see why not, but I’m not entirely sure whether you’re saying we should not support that flag in C++ (neither the positive nor the negative variant) at all or just not in C++11 and later.

File: clang/lib/Format/Format.cpp
Comment (owenca) [inline clang/lib/Format/Format.cpp:3853]: As clang-format formats all C code as C++, there's no need to set the new `RawStringLiterals` option here.

Comment (Sirraide) [reply clang/lib/Format/Format.cpp:3853]: A bunch of the clang-format tests were failing before I made this change, but I think that may have been before I made some more changes to this; I’ll take another look at this.

Comment (owenca) [reply clang/lib/Format/Format.cpp:3853]: Can we set the default to 1 in `LangOpts.def`? That would take care of it.

Comment (Sirraide) [reply clang/lib/Format/Format.cpp:3853]: That sounds like it might work because it’s overridden by the standard anyway; I’ll try that.

**Comment by Sirraide** [comment]
> I think we should allow users to enable them in C++03 modes if -fraw-string-literals is passed. I think it's fine to have -fno-raw-string-literals that allows users to disable them in C++03 mode (in case an earlier command line option opted into them and the user wants to disable the feature for some reason), but I don't know if that's worth the effort to support because I don't think we should allow -fno-raw-string-literals in C++11 mode.

In that case I think it might just make sense to ignore the flag in C++11 and later then and allow it before C++11.

> Btw, it seems that precommit CI found some valid issues to be addressed

Ah, I didn’t know that all new warnings should have a `-W` flag associated with them; it seems I’ve somehow only ever added errors so far. I’ll take a look at what happened there.

File: clang/lib/Format/Format.cpp
Comment (Sirraide) [reply clang/lib/Format/Format.cpp:3853]: That seems to have worked

**Comment by Sirraide** [comment]
@AaronBallman I just noticed something that I’ve somehow not realised until now even though I’d already written a test case for it: Not only does GCC allow raw string literals in gnuXY mode, but also UTF string literals, e.g. `u"foo"` (https://godbolt.org/z/771s8ne5d). 

Should we follow suit here? And if so, should we add a separate flag for that or rename `fraw-string-literals` (and the LangOption) to `fext-string-literals` or something similar?

**Comment by AaronBallman** [comment]
C has Unicode string literals as well: https://godbolt.org/z/chdjYrK9v and so if we're allowing raw string literals, it makes sense to also allow raw unicode string literals IMO. I don't think we need to rename the flag though.

**Comment by AaronBallman** [comment]
> In that case I think it might just make sense to ignore the flag in C++11 and later then and allow it before C++11.

I think that makes the most sense.

File: clang/lib/Format/Format.cpp
Comment (owenca) [reply clang/lib/Format/Format.cpp:3853]: So can undo the change to Format.cpp?

Comment (Sirraide) [reply clang/lib/Format/Format.cpp:3853]: I think I’ve already done that; I just haven’t commited that yet iirc.

**Comment by cor3ntin** [comment]
> C has Unicode string literals as well: https://godbolt.org/z/chdjYrK9v and so if we're allowing raw string literals, it makes sense to also allow raw unicode string literals IMO. I don't think we need to rename the flag though.

Yes, these things are completely orthogonal, it makes no sense to treat raw strings with an encoding prefix differently

**Comment by Sirraide** [comment]
Alright, I think this has the behaviour that we want now:
- raw string literals are enabled in C++11 and later, as well as in C in gnuXY mode;
- raw string literals can be explicitly enabled or disabled in C (and in C++ standards before C++11) using `-f[no-]raw-string-literals`;
- in C++11 and later `-f[no-]raw-string-literals` is ignored and a warning is issued.

**Comment by Sirraide** [comment]
So, apparently, this test here 
https://github.com/llvm/llvm-project/blob/cb76896d6e45e2c9b7ef5e47b6ec37aeca43f7a8/clang/unittests/Lex/DependencyDirectivesScannerTest.cpp#L586-L589

is now failing, presumably because of this:

https://github.com/llvm/llvm-project/blob/cb76896d6e45e2c9b7ef5e47b6ec37aeca43f7a8/clang/lib/Lex/DependencyDirectivesScanner.cpp#L71-L79

I’m not entirely sure how to fix this candidly. It doesn’t look like unconditionally enabling raw string literals is an option here... This situation reminds me of a similar issue we’re having with `'` in numeric literals (#88896). I’m personally not too familiar with the lexer, but would it be possible to pass through the original lang options here from wherever this is invoked?

**Comment by AaronBallman** [comment]
> raw string literals are enabled in C++11 and later, as well as in C in gnu99 mode and later;

Why gnu99 mode and not gnu89 mode? I see GCC has that behavior, but I'm not certain why.

> I’m not entirely sure how to fix this candidly. It doesn’t look like unconditionally enabling raw string literals is an option here... This situation reminds me of a similar issue we’re having with ' in numeric literals (#88896). I’m personally not too familiar with the lexer, but would it be possible to pass through the original lang options here from wherever this is invoked?

Yeah, it's pretty frustrating that we've found two instances of this in such a short period of time. :-/

That test was added in https://github.com/llvm/llvm-project/commit/ee8ed0b3099e63ba0a18cca42b9cfdf098bc6201 and it seems to be a bit of a drive-by as the author noticed the behavior. Given that dependency scanning is never going to care about raw string literals to begin with (at least that I can think of), I'm not certain there's any harm in always supporting raw string literals from dependency scanning, so we could probably do that in the worst case.

But my concerns from https://github.com/llvm/llvm-project/pull/93753#issuecomment-2173666602 are still relevant too. CC @jansvoboda11

**Comment by Sirraide** [comment]
> Why gnu99 mode and not gnu89 mode? I see GCC has that behavior, but I'm not certain why.

We went over this a while back: https://github.com/llvm/llvm-project/pull/88265#discussion_r1571088366

> I'm not certain there's any harm in always supporting raw string literals from dependency scanning, so we could probably do that in the worst case.

But my concerns from https://github.com/llvm/llvm-project/pull/93753#issuecomment-2173666602 are still relevant too. CC @jansvoboda11

:+1:

**Comment by AaronBallman** [comment]
> > Why gnu99 mode and not gnu89 mode? I see GCC has that behavior, but I'm not certain why.
> 
> We went over this a while back: [#88265 (comment)](https://github.com/llvm/llvm-project/pull/88265#discussion_r1571088366)

THAT is why this was so familiar to me! :-D Thanks!

**Review by AaronBallman — APPROVED** [review]
Clang changes LGTM modulo the dependency scanner bits.

**Comment by Sirraide** [comment]
> Clang changes LGTM modulo the dependency scanner bits.

Alright, I’ll wait for a reply from @jansvoboda11 then

**Comment by jansvoboda11** [comment]
I assume that @benlangmuir added the scanner unit-test to demonstrate the current behavior instead of trying to make sure it's preserved. I think making it so that the test passes (actually handles raw string literals) and updating the FIXME in `DependencyDirectivesScanner.cpp` should be fine. CC @Bigcheese

**Comment by jansvoboda11** [comment]
> To clarify, that means setting the `RawStringLiterals` LangOpt in `DependencyDirectivesScanner.cpp`, right? I’m assuming yes, but I just want to make sure.

Yes, that's what I had in mind 👍

**Comment by Sirraide** [comment]
Alright, I just enabled raw string literals in the dependency scanner by default; barring any further complications, I’ll merge this once CI is done.

**Comment by Sirraide** [comment]
Silly me forgot to *actually update the test* after enabling raw string literals in the dependency scanner, but *now* everything should pass.

**Pull request merged**
